pkg/admission/customresourcevalidation/oauth/validate_requestheader.go

```diff
@@ -33,34 +30,37 @@ func ValidateRequestHeaderIdentityProvider(provider *configv1.RequestHeaderIdent
 		return errs
 	}
 
-	if len(provider.ClientCA.Name) > 0 {
-		errs = append(errs, crvalidation.ValidateConfigMapReference(fieldPath.Child("clientCA"), provider.ClientCA, true)...)
-	} else if len(provider.ClientCommonNames) > 0 {
-		errs = append(errs, field.Invalid(fieldPath.Child("clientCommonNames"), provider.ClientCommonNames, "clientCA must be specified in order to use clientCommonNames"))
-	}
+	errs = append(errs, crvalidation.ValidateConfigMapReference(fieldPath.Child("ca"), provider.ClientCA, true)...)
 
 	if len(provider.Headers) == 0 {
 		errs = append(errs, field.Required(fieldPath.Child("headers"), ""))
 	}
 
+	if len(provider.ChallengeURL) == 0 && len(provider.LoginURL) == 0 {
+		errs = append(errs, field.Required(fieldPath, "at least one of challengeURL or loginURL must be specified"))
+	}
+
 	if len(provider.ChallengeURL) > 0 {
-		url, urlErrs := common.ValidateURL(provider.ChallengeURL, fieldPath.Child("challengeURL"))
+		u, urlErrs := common.ValidateURL(provider.ChallengeURL, fieldPath.Child("challengeURL"))
 		errs = append(errs, urlErrs...)
-		if len(urlErrs) == 0 && !strings.Contains(url.RawQuery, urlToken) && !strings.Contains(url.RawQuery, queryToken) {
-			errs = append(errs,
-				field.Invalid(
-					field.NewPath("challengeURL"),
-					provider.ChallengeURL,
-					fmt.Sprintf("query does not include %q or %q, redirect will not preserve original authorize parameters", urlToken, queryToken),
-				),
-			)
+		if len(urlErrs) == 0 {
+			if !hasParamToken(u) {
+				errs = append(errs,
+					field.Invalid(
+						field.NewPath("challengeURL"),
+						provider.ChallengeURL,
+						fmt.Sprintf("query does not include %q or %q, redirect will not preserve original authorize parameters", urlToken, queryToken),
+					),
+				)
+			}
 		}
 	}
+
 	if len(provider.LoginURL) > 0 {
-		url, urlErrs := common.ValidateURL(provider.LoginURL, fieldPath.Child("loginURL"))
+		u, urlErrs := common.ValidateURL(provider.LoginURL, fieldPath.Child("loginURL"))
 		errs = append(errs, urlErrs...)
 		if len(urlErrs) == 0 {
-			if !strings.Contains(url.RawQuery, urlToken) && !strings.Contains(url.RawQuery, queryToken) {
+			if !hasParamToken(u) {
```

Comment:
Note to self: single `if`

Reply:
Actually, nope, there are multiple checks after this one, keep the structure for challengeURL check as well.